### ISSUE TITLE
UICollectionView NSInternalInconsistencyException

### ISSUE THREAD

**Issue by zwang** [badoo/Chatto#142]
Hi @diegosanchezr 

I encounter this weird crash in `performBatchUpdates` and could not figure out where I did wrong. This exception only happens when I quickly send message to a ChatBot and got replies back from the ChatBot (every message sent will get a reply). However, this does not happens on the Chatto Demo App (I added function to generate random incoming message when I post a message). I have spent a couple of days on this and could not figure it out. So I am posting it here to see you can help me or give me some tips how to figure it out when you have some time. 

**Note the numbers, it is always the number after update is 1 greater than that before the updates.** 

Below are all the related code I have  for enqueueModelUpdates. It is very simple and the datasource is also similar to what's in the Demo App. 

Thank you very much for your help. :)

```
   func appendMessage(msg: Message) {
        if isViewVisible() {
            runInSerialBGQueueNow({
                self.dataSource.addMessage(msg)
                self.enqueueModelUpdate(updateType: .Normal)
            })
        }
    }

    func updateMessageStatus(msgId: UInt64, newStatus: MsgStatus) {
        if isViewVisible() {
            runInSerialBGQueueNow({
                self.dataSource.updateMessageStatus(msgId, newStatus: newStatus)
                self.enqueueModelUpdate(updateType: .Normal)
            })
        }
    }
```
> 
> 2016-05-31 17:31:39.943 Pla[1442:545375] *** Assertion failure in -[UICollectionView _endItemAnimationsWithInvalidationContext:tentativelyForReordering:], /BuildRoot/Library/Caches/com.apple.xbs/Sources/UIKit/UIKit-3512.60.12/UICollectionView.m:4625
> 2016-05-31 17:31:45.865 Pla[1442:545375] CRASH: Invalid update: invalid number of items in section 0.  The number of items contained in an existing section after the update (43) must be equal to the number of items contained in that section before the update (42), plus or minus the number of items inserted or deleted from that section (0 inserted, 0 deleted) and plus or minus the number of items moved into or out of that section (0 moved in, 0 moved out).
> 2016-05-31 17:31:45.895 Pla[1442:545375] Stack Trace: (
>     "0   CoreFoundation                      0x00000001828a2dc8 <redacted> + 148",
>     "1   libobjc.A.dylib                     0x0000000181f07f80 objc_exception_throw + 56",
>     "2   CoreFoundation                      0x00000001828a2c80 <redacted> + 0",
>     "3   Foundation                          0x0000000183228154 <redacted> + 112",
>     "4   UIKit                               0x000000018824f00c <redacted> + 12388",
>     "5   UIKit                               0x0000000188252464 <redacted> + 360",
>     "6   UIKit                               0x00000001882522e0 <redacted> + 84",
>     "7   UIKit                               0x0000000187bf02a4 <redacted> + 64",
>     "8   Chatto                              0x0000000101129480 _TFFC6Chatto22BaseChatViewController19performBatchUpdatesFT18updateModelClosureFT_T_7changesVS_17CollectionChanges10updateTypeOS_10UpdateType10completionFT_T__T_U0_FT_T_ + 444",
>     "9   UIKit                               0x0000000187a38fa8 <redacted> + 620",
>     "10  UIKit                               0x0000000187a38d30 <redacted> + 60",
>     "11  Chatto                              0x0000000101127b5c _TTSf4g_s_n_n_n___TFC6Chatto22BaseChatViewController19performBatchUpdatesfT18updateModelClosureFT_T_7changesVS_17CollectionChanges10updateTypeOS_10UpdateType10completionFT_T__T_ + 1172",
>     "12  Chatto                              0x000000010112992c _TFFC6Chatto22BaseChatViewControllerP33_612DFD8221191D415C81FF197E01636612updateModelsFT8newItemsGSaPS_16ChatItemProtocol__8oldItemsGVS_25ReadOnlyOrderedDictionaryVS_17ChatItemCompanion_10updateTypeOS_10UpdateType10completionFT_T__T_U_FT7changesVS_17CollectionChanges18updateModelClosureFT_T__T_ + 304",
>     "13  Chatto                              0x00000001011282d8 _TPA__TFFC6Chatto22BaseChatViewControllerP33_612DFD8221191D415C81FF197E01636612updateModelsFT8newItemsGSaPS_16ChatItemProtocol__8oldItemsGVS_25ReadOnlyOrderedDictionaryVS_17ChatItemCompanion_10updateTypeOS_10UpdateType10completionFT_T__T_U_FT7changesVS_17CollectionChanges18updateModelClosureFT_T__T_ + 132",
>     "14  libdispatch.dylib                   0x0000000101f3da7c _dispatch_call_block_and_release + 24",
>     "15  libdispatch.dylib                   0x0000000101f3da3c _dispatch_client_callout + 16",
>     "16  libdispatch.dylib                   0x0000000101f434e4 _dispatch_main_queue_callback_4CF + 2096",
>     "17  CoreFoundation                      0x0000000182858d50 <redacted> + 12",
>     "18  CoreFoundation                      0x0000000182856bb8 <redacted> + 1628",
>     "19  CoreFoundation                      0x0000000182780c50 CFRunLoopRunSpecific + 384",
>     "20  GraphicsServices                    0x0000000184068088 GSEventRunModal + 180",
>     "21  UIKit                               0x0000000187a6a088 UIApplicationMain + 204",
>     "22  Pla                               0x000000010013af40 main + 144",
>     "23  libdyld.dylib                       0x000000018231e8b8 <redacted> + 4"
> )
> 2016-05-31 17:31:45.902 Pla[1442:545375] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Invalid update: invalid number of items in section 0.  The number of items contained in an existing section after the update (43) must be equal to the number of items contained in that section before the update (42), plus or minus the number of items inserted or deleted from that section (0 inserted, 0 deleted) and plus or minus the number of items moved into or out of that section (0 moved in, 0 moved out).'
> *** First throw call stack:
> (0x1828a2db0 0x181f07f80 0x1828a2c80 0x183228154 0x18824f00c 0x188252464 0x1882522e0 0x187bf02a4 0x101129480 0x187a38fa8 0x187a38d30 0x101127b5c 0x10112992c 0x1011282d8 
> ```0x101f3da7c 0x101f3da3c 0x101f434e4 0x182858d50 0x182856bb8 0x182780c50 0x184068088 0x187a6a088 0x10013af40 0x18231e8b8)
> libc++abi.dylib: terminating with uncaught exception of type NSException

**Comment by diegosanchezr** [comment]
It can be that you are inserting two messages with the same uid? In this case, the engine that calculates the changes won't find anything new and that'd explain the inconsistency.

**Comment by zwang** [comment]
Hmm, I thought about this case and checked the code to make sure there is always an id generated from database insertion and the id should be distinct. Let me look into it further to be sure. I will update back.

Thanks for the help. :) 👍

**Comment by diegosanchezr** [comment]
I didn't notice you're notifying the view controller out of the main thread, that's not supported!

**Comment by zwang** [comment]
@diegosanchezr you are absolutely right. the issue is because of the uid not unique. There was some bug in my code related to fetching the messageId after it being saved to database. After I fixed that bug, I could not reproduce the issue (tested 500+ messages, usually after 10-20 messages I could re-produce the bug)

Regarding "notifying the view controller out of the main thread", I was testing to queue the insert message and datasource update events. It did seems to be OK for me. Anyway, I reverted that testing code. 

Thank you very much again for your help. I really appreciate it. You are so insightful and helpful! 👍

**Comment by nalogowiec** [comment]
@lookitsatravis I saw you referenced my issue here, however - as I checked in my case - I do not have the same ids of the messages in my chat app (or maybe I'm checking it in a wrong way - could you give me any hint how should I test it?) - do you know what other issue could cause this error?

**Comment by lookitsatravis** [comment]
Ah, I thought I edited my comment, but I appear to have just deleted it.

It turns out there was a race condition that I'd inadvertently introduced
and didn't notice before my first comment. I put a breakpoint after the
item was added to the list, and then another right before notifying the
delegate that the data had changed, and I was able to inspect my list to
see if the items changed as I expected. What I didn't notice the first time
is that the method which updated the list of messages was getting called
twice in quick succession - before the change set had been processed from
the first call. The effect is that the first call succeeded, and the second
call was actually duplicating the message item (same UID). I modified my
code to fix the duplicate call, and the error went away.
